### PR TITLE
[@types/node] Add several missing members from `crypto` and `crypto.webcrypto`

### PR DESCRIPTION
**Pull request by jhmaster2000**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:
If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: _(See lists below with links to each member's documentation)_
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### This PR adds the following missing members to the `node:crypto` module:
- [x] [`crypto.DiffieHellmanGroup`](https://nodejs.org/docs/latest/api/crypto.html#class-diffiehellmangroup)
- [x] [`crypto.createDiffieHellmanGroup`](https://nodejs.org/docs/latest/api/crypto.html#cryptocreatediffiehellmangroupname)
- [x] [`crypto.setEngine`](https://nodejs.org/docs/latest/api/crypto.html#cryptosetengineengine-flags)
- [x] [`crypto.setFips`](https://nodejs.org/docs/latest/api/crypto.html#cryptosetfipsbool)
- [x] [`crypto.getRandomValues`](https://nodejs.org/docs/latest/api/crypto.html#cryptogetrandomvaluestypedarray)
- [x] [`crypto.subtle`](https://nodejs.org/docs/latest/api/crypto.html#cryptosubtle)
- [x] [`crypto.webcrypto`](https://nodejs.org/docs/latest/api/crypto.html#cryptowebcrypto)
- [x] [`crypto.webcrypto.subtle`](https://nodejs.org/docs/latest/api/webcrypto.html#cryptosubtle)
- [x] [`crypto.webcrypto.getRandomValues`](https://nodejs.org/docs/latest/api/webcrypto.html#cryptogetrandomvaluestypedarray)
- [x] [`crypto.webcrypto.randomUUID`](https://nodejs.org/docs/latest/api/webcrypto.html#cryptorandomuuid)
- [x] [`crypto.webcrypto.Crypto`](https://nodejs.org/docs/latest/api/webcrypto.html#class-crypto) (Type-only)
- [x] [`crypto.webcrypto.SubtleCrypto`](https://nodejs.org/docs/latest/api/webcrypto.html#class-subtlecrypto) (Type-only)
- [x] [`crypto.webcrypto.CryptoKey`](https://nodejs.org/docs/latest/api/webcrypto.html#class-cryptokey)
- [x] [`crypto.webcrypto.CryptoKeyPair`](https://nodejs.org/docs/latest/api/webcrypto.html#class-cryptokeypair) (Type-only)
- [x] [`crypto.webcrypto.subtle.decrypt`](https://nodejs.org/docs/latest/api/webcrypto.html#subtledecryptalgorithm-key-data)
- [x] [`crypto.webcrypto.subtle.deriveBits`](https://nodejs.org/docs/latest/api/webcrypto.html#subtlederivebitsalgorithm-basekey-length)
- [x] [`crypto.webcrypto.subtle.deriveKey`](https://nodejs.org/docs/latest/api/webcrypto.html#subtlederivekeyalgorithm-basekey-derivedkeyalgorithm-extractable-keyusages)
- [x] [`crypto.webcrypto.subtle.digest`](https://nodejs.org/docs/latest/api/webcrypto.html#subtledigestalgorithm-data)
- [x] [`crypto.webcrypto.subtle.encrypt`](https://nodejs.org/docs/latest/api/webcrypto.html#subtleencryptalgorithm-key-data)
- [x] [`crypto.webcrypto.subtle.exportKey`](https://nodejs.org/docs/latest/api/webcrypto.html#subtleexportkeyformat-key)
- [x] [`crypto.webcrypto.subtle.generateKey`](https://nodejs.org/docs/latest/api/webcrypto.html#subtlegeneratekeyalgorithm-extractable-keyusages)
- [x] [`crypto.webcrypto.subtle.importKey`](https://nodejs.org/docs/latest/api/webcrypto.html#subtleimportkeyformat-keydata-algorithm-extractable-keyusages)
- [x] [`crypto.webcrypto.subtle.sign`](https://nodejs.org/docs/latest/api/webcrypto.html#subtlesignalgorithm-key-data)
- [x] [`crypto.webcrypto.subtle.unwrapKey`](https://nodejs.org/docs/latest/api/webcrypto.html#subtleunwrapkeyformat-wrappedkey-unwrappingkey-unwrapalgo-unwrappedkeyalgo-extractable-keyusages)
- [x] [`crypto.webcrypto.subtle.verify`](https://nodejs.org/docs/latest/api/webcrypto.html#subtleverifyalgorithm-key-signature-data)
- [x] [`crypto.webcrypto.subtle.wrapKey`](https://nodejs.org/docs/latest/api/webcrypto.html#subtlewrapkeyformat-key-wrappingkey-wrapalgo)
- [x] All of WebCrypto's [Algorithm parameters](https://nodejs.org/docs/latest/api/webcrypto.html#algorithm-parameters) interfaces, including the [Node.js specific extensions](https://nodejs.org/docs/latest/api/webcrypto.html#nodejs-specific-extensions), exposed as type-only on `crypto.webcrypto.SubtleCrypto.*`

### Other changes
- [x] Changed [`diffieHellman.computeSecret`](https://nodejs.org/docs/latest/api/crypto.html#diffiehellmancomputesecretotherpublickey-inputencoding-outputencoding) overloads to be more strict for further type safety.
- [x] Changed [`crypto.getDiffieHellman`](https://nodejs.org/docs/latest/api/crypto.html#cryptogetdiffiehellmangroupname) return type from `DiffieHellman` to `DiffieHellmanGroup` as it was incorrect.
- [x] A test file (`util_types.ts`) contained a test for `CryptoKey` in which it attempted to call `new webcrypto.CryptoKey()`, however `CryptoKey`'s constructor is illegal to be called and unconditionally throws, it has been typed accordingly so this test no longer passed, it has been changed to obtain a `CryptoKey` the correct way through `webcrypto.subtle.generateKey`
- [x] The `SubtleCrypto` namespace contains many helper types to more easily type it's functions.
- [x] Added all relevant tests to `crypto.ts`
- [x] The tests of the `diffie-hellman` package had to be updated due to incorrectly expecting `DiffieHellman` as the return type of `crypto.getDiffieHellman`, which as stated above was an incorrect typing and has now been corrected to the correct return type of `DiffieHellmanGroup`, so the mentioned package's tests have been updated accordingly.
- [x] Backported relevant typings to versions `v14`, `v16` and `v17`.
- [x] Implemented changes to Web Crypto in `v18.4.0` ([docs](https://nodejs.org/docs/latest-v18.x/api/webcrypto.html#web-crypto-api))
  - [x] Removed proprietary 'node.keyObject' import/export format.
  - [x] Removed proprietary 'NODE-DSA', 'NODE-DH', and 'NODE-SCRYPT' algorithms.
  - [x] Added 'Ed25519', 'Ed448', 'X25519', and 'X448' algorithms.
  - [x] Removed proprietary 'NODE-ED25519' and 'NODE-ED448' algorithms.
  - [x] Removed proprietary 'NODE-X25519' and 'NODE-X448' named curves from the 'ECDH' algorithm.
  - [x] `subtle.deriveBits`: Added 'X25519', and 'X448' algorithms.
  - [x] `subtle.deriveKey`: Added 'X25519', and 'X448' algorithms.
  - [x] `subtle.exportKey`: Added 'Ed25519', 'Ed448', 'X25519', and 'X448' algorithms.
  - [x] `subtle.importKey`: Added 'Ed25519', 'Ed448', 'X25519', and 'X448' algorithms.
  - [x] `subtle.sign`: Added 'Ed25519', and 'Ed448' algorithms.
  - [x] `subtle.verify`: Added 'Ed25519', and 'Ed448' algorithms.
  - [x] Added class: [`AlgorithmIdentifier`](https://nodejs.org/docs/latest-v18.x/api/webcrypto.html#class-algorithmidentifier)
  - [x] Added class: [`Ed448Params`](https://nodejs.org/docs/latest-v18.x/api/webcrypto.html#class-ed448params)
  - [x] Some of the **Algorithm parameters** seem to have been removed without notice, such as [`RsaSignParams`](https://nodejs.org/docs/latest-v17.x/api/webcrypto.html#class-rsasignparams), these have presumably all been merged into the new `AlgorithmIdentifier`.
  - [x] All changes relating to [webcrypto secure curves](https://nodejs.org/docs/latest-v18.x/api/webcrypto.html#user-content-fn-1) based on the linked footnote references.
  - [x] Types are aligned to `@types/web`. (Thanks @panva) 

### Details
This PR is now complete and should be ready to merge.

### CODE REVIEW

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## diffie-hellman ([<kbd>unpkg</kbd>](https://unpkg.com/browse/diffie-hellman@latest/))
was missing the following properties:
1. DiffieHellmanGroup
1. createDiffieHellmanGroup
1. DiffieHellmanGroup
1. createDiffieHellmanGroup
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 0bb80560f9163a1ba5b920b035370e6d07c28f0a
</p>

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by peterblazejewicz** [comment]
can this one wait until #60047 is merged? I'll help with eventual updates/resolutions (just concern about impacting 60047 work)

**Comment by peterblazejewicz** [comment]
The v18 has been merged

**Review by peterblazejewicz — APPROVED** [review]

**Comment by jhmaster2000** [comment]
Alright thank you both for the reviews, I will quickly backport the relevant APIs to the v14 and v16 folders as some of these additions are rather old (not v12 as I assume with the merge of v18 and v12 already being EOL it will be removed soon) sometime later today and then merge.

**Comment by jhmaster2000** [comment]
It would appear Node.js v18.4.0 is out and with it some rather significant changes to their Web Crypto API implementation: https://nodejs.org/docs/latest/api/webcrypto.html#web-crypto-api (Open History dropdown).

I will be including those latest changes on this PR aswell so it may take a little longer than expected.

**Comment by jhmaster2000** [comment]
> > as I am unsure exactly how @types/node typings versioning semantics work, could I be given some guidance regarding this?
> 
> https://github.com/DefinitelyTyped/DefinitelyTyped/#how-do-definitely-typed-package-versions-relate-to-versions-of-the-corresponding-library
> 
> For non-current (v18) changes, relevant changes should be added to respective modules within their versioned folders (say: /v16/...., etc). That's all required here for versioning, no changes to configuration

Oh so the `"If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header."` checkbox on the template does not apply to `@types/node`? I assume it's handled automatically by a bot or script then

**Comment by jakebailey** [comment]
I would assume that if a feature is only available in node 18.4, that the minor version in the types package is also updated to that point release; that's definitely how it works for "normal" packages, and I think that applies to node packages like how a PR like this bumped it: https://github.com/DefinitelyTyped/DefinitelyTyped/commit/cd10dd77c79eb9b3c987edadd5eb295b098a9add#diff-a477538df71d810bc99f9f5861c0635bd85db9a1c32bd2b72b68e3cd9437a4e8R1

**Comment by jhmaster2000** [comment]
Upon reading more carefully through DT's README that does seem to be the case for the main folder, only the `v##` folders are exempt as those Node versions no longer received minor updates, only patches, which DT versioning doesn't sync with.

With this in mind I will bump the main Node typings up to `v18.4` together with the 18.4 changes to Web Crypto.

**Review by panva — CHANGES_REQUESTED** [review]
Amazing!

Quick note, i see it's on the TODO.

All Node.js-specific extensions (NODE-DH, NODE-SCRYPT, NODE-DSA, NODE-DH, NODE-ED25519, NODE-ED448, node.keyObject, NODE-X25519, and NODE-X448) have been removed in 18.4.0.

The webcrypto types and interfaces should reflect how these are in `@types/web` (e.g. no need for dictionaries that are covered by AlgorithmIdentifier) with one or two exceptions, Buffer is also a valid `BufferSource` and the addition of [webcrypto secure curves](https://github.com/WICG/webcrypto-secure-curves) adds an additional dictionary, `Ed448Params`, as valid sign/verify algorithm.

**Comment by jhmaster2000** [comment]
> Amazing!
> 
> Quick note, i see it's on the TODO.
> 
> All Node.js-specific extensions (NODE-DH, NODE-SCRYPT, NODE-DSA, NODE-DH, NODE-ED25519, NODE-ED448, node.keyObject, NODE-X25519, and NODE-X448) have been removed in 18.4.0.
> 
> The webcrypto types and interfaces should reflect how these are in `@types/web` (e.g. no need for dictionaries that are covered by AlgorithmIdentifier) with one or two exceptions, Buffer is also a valid `BufferSource` and the addition of [webcrypto secure curves](https://github.com/WICG/webcrypto-secure-curves) adds an additional dictionary, `Ed448Params`, as valid sign/verify algorithm.

Yes as stated those are already in the TODO, near the bottom, I specify the addition of AlgorithmIdentifier and the removal of some interfaces in favor of it. Thanks for pointing me to `@types/web` I will be sure to use it as reference from now on.

Regarding `Buffer`, it extends `Uint8Array`, so it is already assignable to web TypedArray type such as `BufferSource` without needing to explicitly add `Buffer` on the type. It seems the `node:stream/web` module already declares a `BufferSource` for the Node context, which is identical to the web one, so I will just import that instead of redeclaring it on crypto.d.ts

**Comment by jhmaster2000** [comment]
Well this doesn't look good, I'm not sure why this happened, it looks fine from here: https://github.com/DefinitelyTyped/DefinitelyTyped/compare/master...jhmaster2000:master 😕

File: types/node/v16/crypto.d.ts
Comment (domdomegg) [inline types/node/v16/crypto.d.ts:3752]: This name is not very obvious and I initially read it as `RsaOaepParams`. Would `EncryptionAlgorithmParams` prevent this and also allow for easier extension e.g. if it introduces quantum safe crypto encryption algorithms.

**Review by domdomegg — APPROVED** [review]
These definitions are excellent, thank you so much @jhmaster2000 for writing them and others for supporting and reviewing.

I've added a couple minor renaming suggestions but think these are GTG.

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by panva** [comment]
@jhmaster2000 do you plan on picking up the remaining TODOs? If not I can pick this PR up and align the interfaces with `@types/web`

**Review by panva — CHANGES_REQUESTED** [review]
Aside of the remaining TODO there are incorrect types wrt to how WebCryptoAPI CryptoKey usages, extractable, etc actually works at times, I'll attempt open a PR against against @jhmaster2000's fork with fixes.

**Comment by jhmaster2000** [comment]
> @jhmaster2000 do you plan on picking up the remaining TODOs? If not I can pick this PR up and align the interfaces with `@types/web` myself.

I have not forgotten or abandoned this PR, but I havs found myself rather busy this past week in combination with having gotten ill for the last few days. Nonetheless I do welcome the help and contribution! So please feel free to pick up any TODOs if you wish.
I expect to be able to catch up with this PR on the weekend where I'll be able to review and merge your PR on my branch.

**Review by panva — APPROVED** [review]

File: types/node/v16/crypto.d.ts
Comment (jhmaster2000) [reply types/node/v16/crypto.d.ts]: Resolved by this PR: https://github.com/jhmaster2000/DefinitelyTyped/pull/2

File: types/node/v16/crypto.d.ts
Comment (jhmaster2000) [reply types/node/v16/crypto.d.ts]: Resolved by this PR: https://github.com/jhmaster2000/DefinitelyTyped/pull/2

File: types/node/v16/crypto.d.ts
Comment (jhmaster2000) [reply types/node/v16/crypto.d.ts]: Resolved by this PR: https://github.com/jhmaster2000/DefinitelyTyped/pull/2

**Comment by jhmaster2000** [comment]
PR should now be ready to merge for good this time, just need to await reviews now

**Review by panva — APPROVED** [review]

**Review by domdomegg — APPROVED** [review]
Thanks @jhmaster2000 and @panva for your great work on this!

**Review by panva — APPROVED** [review]

**Review by mcollina — APPROVED** [review]
lgtm

**Review by peterblazejewicz — APPROVED** [review]

**Comment by panva** [comment]
Follow up fix to `CryptoKey.usages` in #61240

**Pull request merged**
